compiler/rustc_hir_typeck/src/fn_ctxt/checks.rs

```diff
@@ -1955,6 +1964,410 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+
+                        // Use the sole variant definition as our variant from now on:
+                        (struct_variant_def, struct_def_generics)
+                    };
+
+                    // Only retain the generics which are relevant (according to the `impl`).
+                    let struct_def_generics: Vec<&ty::GenericParamDef> = relevant_ty_args_indices
+                        .into_iter()
+                        .filter_map(|index| struct_def_generics.params.get(index))
+                        .collect();
+
+                    // The struct being constructed is the same as the struct in the impl.
+
+                    let blameable_field_defs: Vec<&ty::FieldDef> = struct_variant_def
+                        .fields
+                        .iter()
+                        .filter(|field_def| {
+                            let field_type: Ty<'tcx> = self.tcx.type_of(field_def.did);
+
+                            // Only retain fields which mention the blameable generics.
+                            struct_def_generics
+                                .iter()
+                                .any(|param| find_param_def_in_ty_walker(field_type.walk(), param))
+                        })
+                        .collect();
+
+                    if blameable_field_defs.len() != 1 {
+                        // We must have a single blameable field, in order to be able to blame it.
+                        return Err(expr);
+                    }
+
+                    let blameable_field_def: &ty::FieldDef = blameable_field_defs[0];
+                    let blameable_field_index = struct_variant_def
+                        .fields
+                        .iter()
+                        .position(|field_def| field_def == blameable_field_def)
+                        .expect("exists in list");
+
+                    if blameable_field_index < call_ctor_args.len() {
+                        return Ok(&call_ctor_args[blameable_field_index]);
+                    }
+
+                    // We failed to find a matching field in the original struct expression.
+                    Err(expr)
+                }
+                _ => Err(expr),
+            },
+            hir::ExprKind::Tup(elements) => {
+                if let &[single_index] = relevant_ty_args_indices.as_slice() {
+                    if single_index < elements.len() {
+                        return Ok(&elements[single_index]);
+                    }
+                }
+                return Err(expr);
+            }
+            _ => Err(expr),
+        }
+    }
+
+    fn point_at_specific_expr_if_possible_for_predicate_obligation(
+        &self,
+        obligation: &PredicateObligation<'tcx>,
+        expr: &'tcx hir::Expr<'tcx>,
+    ) -> Result<&'tcx hir::Expr<'tcx>, &'tcx hir::Expr<'tcx>> {
+        self.point_at_specific_expr_if_possible_for_obligation_cause_code(
+            obligation.cause.code(),
+            expr,
+        )
+    }
```

Comment:
Why is this a separate function? Can you inline its body into all of the callers?

Reply:
I'm also thinking if these functions could get more descriptive names -- perhaps `find_best_blame_expr_for_obligation`?

Reply:
Originally I thought I was going to need more functions, so there was supposed to be some sort of hierarchy where they called each other to handle different types of constraints/expressions.

It turns out not to work that way, so this is probably superfluous.